### ISSUE TITLE
Assignment 3 - Parsing Input

### ISSUE THREAD

**Issue by DanTheTingler** [mikeizbicki/ucr-cs100#65]
So for assignment 1 I used the function strtok to parse the input and eliminate any spaces. For assignment 3 when we parse input, we should be looking for the following characters: <, >, >>, |   correct? Then we have to remove them so we can run the commands on evec right? Can anyone can give me a hint as to what way would be the best way to parse my input? I'm rather stuck on this portion.

**Comment by bmars003** [comment]
Correct. You will need to check the input from the user for `<, >, >>, |` and if they show up then you will need to pipe accordingly.

**Comment by DanTheTingler** [comment]
Hmm. So I did this to try to check for user input:

 for(int p  = 0; p < 1024; p++)
        {
            if(str[p] == '#') //if so, set the # to NULL, effectively igorning everything after
            {
                str[p] = '\0';
            }

```
        if(str[p] == '<')
        {
            bcarrot = true;
        }

        if(str[p] == '>')
        {
            if(str[p+1] == '>')
            {
                dcarrot = true;
                fcarrot = false;
            }

            else
            {
                fcarrot = true;
            }
        }

        if(str[p] == '|')
        {
            pole = true;
        }
    }
```

And for some reason everytime I enter ">>" both fcarrot and dcarrot end up both being true. How come?

**Comment by DanTheTingler** [comment]
Okay the formatting for my post just got a bit weird, everything that I just posted is inside the for loop.

**Comment by bmars003** [comment]
I think about how you are checking for `>>`. 
You first check if `>` shows up if so then you check to see if `>` is the preceeding character . If both the preecedding character is a `>` then you set `dcarrot` = `true` and `fcarrot` = `false`. What happens if the next character I look at is the second `>`. What does your if statement check? It check to see if the preeceeding character is a `>`, which in most cases won't happenen unless  the user enters `>>>`. So the 
`
if(str[p+1] == '>')
`
statemenet will return `false`, so the program goes to the `else statement` , where you set `fcarrot` = `true`.

Does this make sense?

**Comment by DanTheTingler** [comment]
Since I'm looking at p+1, isn't that the succeeding character after p?

**Comment by bmars003** [comment]
Yes, but look at what happens when you look at the second '>'. If you are on campus come to my OH and I can explain. See post https://github.com/mikeizbicki/ucr-cs100/issues/63 .

**Comment by DanTheTingler** [comment]
I commute so unfortunately I'm not on campus. I still don't get it though, my logic is if the first character is > and the one following is also >, then dcarrot is true. If the succeeding character isn't, then only fcarrot is true.

**Comment by bmars003** [comment]
You are getting stuck if str[p] is the first `>`. Look at your logic when str[p] is the second '>'. You will notice that str[p] is indeed `>`, but the preeceeding character( str[p+1] ) will not be a '>', therefore the if statement is NOT taken and the else statement is taken resulting in `fcarrot` = `true`.

**Comment by DanTheTingler** [comment]
Hmm okay. So what would be the best to fix this? I don't know how to check for >> since it can't be put in single quotes.

**Comment by bmars003** [comment]
My question to you, is what is the purpose of the for loop?

**Comment by DanTheTingler** [comment]
It's checking to see if the user has entered those symbols so I can run the necessary functions.

**Comment by DanTheTingler** [comment]
Only one I'm having trouble with is the >>

**Comment by bmars003** [comment]
Would your program be able to handle chaining of I/O redirection?

**Comment by DanTheTingler** [comment]
Hmm I don't think it would, back to square one. What would be the best way to parse the input and recognize that the user is inputting these symbols?

**Comment by bmars003** [comment]
I would  work out a way that parses the input from the user in a way that is able to separate commands that needs to be run, any I/O redirection if any and `&`. I would start parsing from left to right and place each parsed component into  an array or other container.

**Comment by DanTheTingler** [comment]
Okay the only symbol I'm having trouble with is still the >>. I followed what you said, and instead of setting boolean variables to true, I'm pushing them into a vector. However, my vector is a vector of chars, and ">>" is not a char, so what do I do?

**Comment by mikeizbicki** [comment]
Obviously it shouldn't be a vector a chars...

**Comment by DanTheTingler** [comment]
Ok right, I fixed that, I just wanted it to be a vector of chars since I was using cstrings. Is there any other function I can use to parse my input besides strtok? Current that's what I'm using, and it parses <, >, and | fine, but it can't read in >>...

**Comment by DanTheTingler** [comment]
Okay nvm, I got strktok to work. I just can't seem to get my if statements right - if I enter >>, it always says I entered > as well as >>. Here's what I have for checking for the both of them:

for(int p  = 0; p < 1024; p++)
{
             if(str[p] == '#') //if so, set the # to NULL, effectively igorning everything after
            {
                str[p] = '\0';
            }

```
        if(str[p] == '<')
        {
            bcarrot = true;
            commands.push_back("<");
        }

        if(str[p] == '>' && str[p+1] != '>')
        {
            fcarrot = true;
            commands.push_back(">");
        }

        if(str[p] == '>' && str[p+1] == '>')
        {
            dcarrot = true;
            commands.push_back(">>");
        }

        if(str[p] == '|')
        {
            pole = true;
            commands.push_back("|");
        }
```

}

**Comment by DanTheTingler** [comment]
I know why I'm getting it though. When I enter >>, it sets dcarrot to true, then iterates on. On the next iteration however, it is now looking at the second >, which sets > to true. How do I avoid this problem without converting everything to strings?

**Comment by DanTheTingler** [comment]
Okay I got it, thanks for the help Bryan and Mike! Let me know if you want me to post how I solved it.

**Comment by bmars003** [comment]
It might be helpful to others if you did.

**Comment by DanTheTingler** [comment]
Sounds good! So what I did was when it was >>, in the if statement for it, after I pushed into the vector, I set str[p+1] to a ' ', so that on the next iteration of the for loop there wouldn't be that problem. And since I pushed back into a vector, everything was recorded in the exact order. Even if a letter were to directly succeed the >>, this wouldn't be a problem, as it would only change the >> to a >' ', which doesn't affect what comes after.
